### PR TITLE
Include Modal on not-signed-in Homepage of a program

### PR DESCRIPTION
**Pull request by innopoop**
## Notes
1. AuthModal Component is used individually when a program name is involved (/program/[slug] or /apply). Otherwise, with no program name, it pops up on top of what the page would look like while logged in (see "My Profile Page")
2. Sign up and login routes operate as they did before
3. If logged in and you go to /login or /signup, redirects to home
4. Deprecated RedirectIfNotLoggedIn
5. Integrated AuthModal into AuthorizationWrapper for non-program related sign in instances

TODO: Deprecate /my/applications for display of applications on homepage

## Behavior
Program Homepage
1. Not logged in -> Click on apply as mentor/mentee -> opens authentication modal
2. Signed in but unverified -> Click on apply as mentor/mentee -> opens authentication modal
3. Sign up with email after authentication modal pops up -> takes you to verify with to=program home
4. Log in / sign up successfully -> refreshes page content and allows user to navigate to apply route
image

Apply Page
1. Not logged in / Unverified
image
2. Logged in
image

My Profile Page
image

Create Program Page
image

### CODE REVIEW

**Comment by nova-admin** [comment]
#### 📝Changed routes:


| [**/create**](https://risc-gztuh6hha-novaforgood.vercel.app/create) | [**/**](https://risc-gztuh6hha-novaforgood.vercel.app/) |
|:-:|:-:|
|<a href="https://risc-gztuh6hha-novaforgood.vercel.app/create">Screenshot of /create</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/oUnOAHPhEHPrzf2Rv4HHAjpRg9EtqMG7dbBFFe9qe4HM56q9u3QOgrYUE9-1_XhMovuWnjK0p5T8_S9p10wIxw.png">(view full size)</a> |<a href="https://risc-gztuh6hha-novaforgood.vercel.app/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/DN-sBIK7owATnPn9tTV_DfEL74zefP7C-4gHGSqZO2sv-FTyU4C8iOEm8V5n5Azd4tjIL2ldP_WNVq8lowQlaw.png">(view full size)</a> |



| [**/program/\[slug\]/apply**](https://risc-gztuh6hha-novaforgood.vercel.app/program/[slug]/apply) | [**/program/\[slug\]**](https://risc-gztuh6hha-novaforgood.vercel.app/program/[slug]) |
|:-:|:-:|
|<a href="https://risc-gztuh6hha-novaforgood.vercel.app/program/[slug]/apply">Screenshot of /program/[slug]/apply</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/exM9wjcbaGujiq2OtWo2n7Xva3yWUFI9j_KGhmZlidjz9qgtwUwXtlET7ej7aB1xRZ0oZf8LPNO_P98ovDMykA.png">(view full size)</a> |<a href="https://risc-gztuh6hha-novaforgood.vercel.app/program/[slug]">Screenshot of /program/[slug]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/DnpnYSliRaE11H-z2iSFLvAdZ699ue7SixAOCOOtmVJKTo4esRcCdRTkjYObgRlfWfOhq0wvNVShl1EI9iVG5g.png">(view full size)</a> |



| [**/reset-password**](https://risc-gztuh6hha-novaforgood.vercel.app/reset-password) | [**/signup**](https://risc-gztuh6hha-novaforgood.vercel.app/signup) |
|:-:|:-:|
|<a href="https://risc-gztuh6hha-novaforgood.vercel.app/reset-password">Screenshot of /reset-password</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/zuN9DWALaftfzq6JY6b2e7h1ME5ezSlA0elzb78uzVDYpdakswFbUtTb2WV3IUEFsf8A-tIOxxeYLGc8P_BJ9w.png">(view full size)</a> |<a href="https://risc-gztuh6hha-novaforgood.vercel.app/signup">Screenshot of /signup</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/jQUj_HAJ3fLJag6OpfMnkkTKZ3rdmaLbcSoa1TmfKxUi8LhFhW8g3WZWd6haDb4c0zHxgTZjY4c7X0VvCZscGQ.png">(view full size)</a> |







Commit 530c03d7387747e0c4c2457f0995ab25edd971bc (https://risc-gztuh6hha-novaforgood.vercel.app).

File: src/components/Authentication/AuthModal.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+          className="cursor-pointer focus:outline-none self-end"
+        >
+          <img src="/static/Close.svg" className="h-4 w-4" />
+        </button>
+        {isLogin ? (
+          <div>
+            <div className="flex flex-col space-y-1">
+              {programName ? (
+                <Text h3 b>
+                  Login to join {programName}
```
Comment (legitmaxwu) [inline src/components/Authentication/AuthModal.tsx:31]: Maybe use a different color to make `programName` more distinguishable?

File: src/components/Authentication/Login.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+          onClick={(e) => {
+            e.preventDefault();
+            setLoading(true);
+            signInWithEmail(email, password)
+              .then((res) => {
+                if (res) {
+                  if (!res.user?.emailVerified) {
+                    setError("Please verify this email before logging in.");
+                  } else {
+                    onSuccessfulEmailLogin();
```
Comment (legitmaxwu) [inline src/components/Authentication/Login.tsx:95]: nit: I feel like it's clearer to just move it out of the `else` block, but up to u if you wanna change it.

File: src/layouts/AuthorizationWrapper.tsx [hunk 3]
```diff
@@ -2,6 +2,7 @@ import { useRouter } from "next/router";
 import React, { Fragment } from "react";
 import ErrorScreen, { ErrorScreenType } from "../components/ErrorScreen";
 import { AuthorizationLevel, useAuthorizationLevel } from "../hooks";
+import AuthModal from "../components/Authentication/AuthModal";
```
Comment (legitmaxwu) [inline src/layouts/AuthorizationWrapper.tsx:5]: nit: I'd call this `AuthenticationModal` to avoid potential confusion with "Authorization"

File: src/pages/create.tsx [hunk 4]
```diff
@@ -143,9 +142,6 @@ const CreateProgramPage: Page = () => {
         if (err.message.includes("duplicate key value violates unique"))
           setError(`Error: Slug "${programIdentifier}" already exists`);
         else setError("Error: " + err.message);
-      })
-      .finally(() => {
-        setLoading(false);
```
Comment (legitmaxwu) [inline src/pages/create.tsx:148]: Why is this removed? If there's an error on creating program, would it still be possible to create a program upon fixing the error?

File: src/pages/create.tsx [hunk 5]
```diff
@@ -211,54 +220,58 @@ const CreateProgramPage: Page = () => {
             onSrcChange={setProgramLogoURL}
           />
         </div>
+        <form className="space-y-10">
```
Comment (legitmaxwu) [inline src/pages/create.tsx:223]: nit: Personally I'd favor using `div` spacers over `space-y-10` but up to u

File: src/pages/create.tsx [hunk 6]
```diff
@@ -160,30 +156,43 @@ const CreateProgramPage: Page = () => {
...
-          onClick={() => {
-            if (validateProgramName(programName)) {
-              setStage((prev) => prev + 1);
-              setError("");
-            } else setError("Please enter a program name.");
-          }}
-        >
-          Next
-        </Button>
+        <form className="space-y-6">
```
Comment (legitmaxwu) [inline src/pages/create.tsx:159]: nit: Personally I'd favor using `div` spacers over `space-y` but up to u

File: src/components/Authentication/AuthModal.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                </Text>
+              ) : (
+                <Text h3 b>
+                  Login to Access the Mentor Center
```
Comment (legitmaxwu) [inline src/components/Authentication/AuthModal.tsx:35]: "Login to access Mentor Center" imo? same w/signup

File: src/pages/create.tsx [hunk 8]
```diff
@@ -211,54 +220,58 @@ const CreateProgramPage: Page = () => {
...
+                e.preventDefault();
+                if (!validateProgramIdentifier(programIdentifier))
+                  setError(
+                    "Your program identifier must be alphanumeric and have at least 4 characters."
+                  );
+                else if (
+                  !loading &&
+                  validateProgramIdentifier(programIdentifier)
+                ) {
+                  setError("");
```
Comment (legitmaxwu) [inline src/pages/create.tsx:266]: `setError(null)` might be more clear (and then `useState<string | null>`)

File: src/pages/create.tsx [hunk 9]
```diff
@@ -211,54 +220,58 @@ const CreateProgramPage: Page = () => {
...
+              type="submit"
+              disabled={loading || programIdentifier.length == 0}
+              onClick={(e) => {
+                e.preventDefault();
+                if (!validateProgramIdentifier(programIdentifier))
+                  setError(
+                    "Your program identifier must be alphanumeric and have at least 4 characters."
+                  );
+                else if (
+                  !loading &&
```
Comment (legitmaxwu) [inline src/pages/create.tsx:263]: I think the `loading` check should be moved tot he beginning, i.e. `if (loading) return`

File: src/pages/signup.tsx [hunk 10]
```diff
@@ -19,21 +18,18 @@ const BlobCircle = () => {
...
-  // const [tocChecked, setTocChecked] = useState(false);
+  const authorizationLevel = useAuthorizationLevel();
   const router = useRouter();
 
+  if (
+    authorizationLevel !== AuthorizationLevel.Unverified &&
+    authorizationLevel !== AuthorizationLevel.Unauthenticated
+  ) {
+    router.push("/");
+    return <div>Logged in. Redirecting...</div>;
```
Comment (legitmaxwu) [inline src/pages/signup.tsx:30]: Maybe return a fragment or some other loading screen (text is jank)

(Same for login)

File: tailwind.config.js [hunk 11]
```diff
@@ -70,6 +70,7 @@ module.exports = {
       },
       width: {
         "9/10": "90%",
+        "36rem": "36rem",
```
Comment (legitmaxwu) [inline tailwind.config.js:73]: Why not follow normal sizing conventions?

i.e. "120": "30rem"

File: src/layouts/SignedInAsWrapper.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import { Text } from "../components/atomic";
+import { useGetMyUserQuery } from "../generated/graphql";
+import { AuthorizationLevel, useAuthorizationLevel } from "../hooks";
+import { useAuth } from "../utils/firebase/auth";
+
+const SignedInAsWrapper: React.FC = ({ children }) => {
```
Comment (legitmaxwu) [inline src/layouts/SignedInAsWrapper.tsx:6]: This doesn't seem like it needs to be a wrapper--consider making it a component?

**Review by legitmaxwu — CHANGES_REQUESTED** [review]
Good work! I love how you abstracted `Login` and `Signup` into their own components, and made a bunch of fixes throughout the codebase.

One question
- When I open the modal, click to the Login screen, close the modal, and then open it again, I'm still on the Login screen. Is that intentional or should it always default to show Signup?

Other than that, left some code comments but it's looking good!

File: src/pages/create.tsx (hunk 4 above)
Comment (innopoop) [reply src/pages/create.tsx:148]: Really good point- I removed it because it wasn't disabling the button after the program creation was successful so perhaps including it in the .catch statement and leaving it as loading on successful creation would be the move?

Comment (legitmaxwu) [reply src/pages/create.tsx:148]: sg, including it in the catch statement is sensible to me!

File: src/pages/create.tsx (hunk 5 above)
Comment (innopoop) [reply src/pages/create.tsx:223]: I personally like using space when there are more than like 2 or 3 HTML elements because it reduces the amount of code that is written :0

**Comment by innopoop** [comment]
> Good work! I love how you abstracted `Login` and `Signup` into their own components, and made a bunch of fixes throughout the codebase.
> 
> One question
> 
> * When I open the modal, click to the Login screen, close the modal, and then open it again, I'm still on the Login screen. Is that intentional or should it always default to show Signup?
> 
> Other than that, left some code comments but it's looking good!

I'm not sure if it makes more sense to have it default to sign up o-o If a user clicked log in, wouldn't they want to stay on login? We can discuss this maybe and come back to it.

File: src/components/Authentication/Login.tsx (hunk 2 above)
Comment (innopoop) [reply src/components/Authentication/Login.tsx:95]: For this, did you mean to call `onSuccessfulEmailLogin()` even if email is not verified?

File: src/components/Authentication/AuthModal.tsx (hunk 1 above)
Comment (innopoop) [reply src/components/Authentication/AuthModal.tsx:31]: Would italices be enough for distinguishing program name? Not sure what color we could change it to to make it look nice.
image

Comment (legitmaxwu) [reply src/components/Authentication/AuthModal.tsx]: I feel like italics might be too much, but good point. Idk either

File: src/components/Authentication/Login.tsx (hunk 2 above)
Comment (legitmaxwu) [reply src/components/Authentication/Login.tsx:100]: I was thinking return after setError. It's not a huge deal, but there might be other things that prevent `onSuccessfulEmailLogin` from being called so in my head not using `else` is more scalable

**Comment by innopoop** [comment]
Create

> > Good work! I love how you abstracted `Login` and `Signup` into their own components, and made a bunch of fixes throughout the codebase.
> > One question
> > 
> > * When I open the modal, click to the Login screen, close the modal, and then open it again, I'm still on the Login screen. Is that intentional or should it always default to show Signup?
> > 
> > Other than that, left some code comments but it's looking good!
> 
> Thank you Max!! I'm not sure if it makes more sense to have it default to sign up o-o If a user clicked log in, wouldn't they want to stay on login? We can discuss this maybe and come back to it.

Make it default to create!

**Review by legitmaxwu — APPROVED** [review]
